Feature added : Scanning of directories (Issue #77)

SinghShreya05 opened the pull request:
1. Made changes in atarashii.py file
2. Added a "scandir" function
3. Added a "--scandir"  in ArgumentParser

This is with respect to issue #77. Please review the PR. @hastagAB @Kaushl2208

GMishx commented on atarashi/atarashii.py:
```diff
@@ -107,18 +120,29 @@ def main():
+
+  if scandirfilepath:
+        try:
+          filestorejson = scandir(scandirfilepath)
```
Hello @SinghShreya05, thank you for your contribution.
I see you are getting the list of files inside a directory but nothing is done after that. Are you planning to push some more code here?

GMishx commented on atarashi/atarashii.py:
```diff
@@ -107,18 +120,29 @@ def main():
                       help="Specify the similarity algorithm that you want."
                            " First 2 are for TFIDF and last 3 are for Ngram")
+
+  parser.add_argument("-d", "--scandir", help="Extracting all files from directories and subdirectories. Returns everything in JSON")
```
Adding a `--scandir` flag to the list is fine but it will still left user to provide a `inputFile` if they want to scan a directory and vice-versa. Will it be possible to read the directory from `inputFile` itself and use `os.path.isdir()` to differentiate?

SinghShreya05 replied:
I understood my error. Instead of adding a new argument "scandir" I should extract all the files using "inputFile" argument passed on the basis of whether it's a file or a directory, if it's a directory all files will be scanned and for each file "atarashii_runner" will be called. Also, each result will be stored in JSON. Am I proceeding in the right direction now?

GMishx replied:
Yes, that will be more aligned to the issue.

SinghShreya05 commented:
1. Checked whether "inputFile" is either a file or a directory.
2. If it's a directory, every file inside the directory or sub-directories will be scanned and for each file "atarashii_runner" will be called. 
3. Each result will be printed as they come and will be stored in JSON.
4. Every change is made in the "atarashii.py" file itself.
This is with respect to issue #77.

SinghShreya05 commented:
1. Checked whether "inputFile" is either a file or a directory.
2. If it's a directory, every file inside the directory or sub-directories will be scanned and for each file "atarashii_runner" will be called.
3. Each result will be printed as they come and will be stored in JSON.
4. Every change is made in the "atarashii.py" file itself.

This is with respect to issue #77

GMishx commented on atarashi/atarashii.py:
```diff
@@ -126,28 +126,57 @@ def main():
+                              "description": ""
+                              }]
+                      result = list(result)
+                      dirresult[fpath] = result
+                      result = {"file": fpath, "results": result}
+                      print(result + "\n")
+          dirresult = json.dumps(dirresult, sort_keys=True, ensure_ascii=False, indent=4)                
```
1. It will be more beneficial if the modified result (dictionary with `file` and `results`) be added to `dirresult` as that is the final result for a single file scan.
2. Plain print here does not help much. If you can do something clever, like printing every result as it arrives while maintaining the valid JSON format, will be more desirable as the user will not have to wait to see if atarashi is doing something.
3. The final `json.dumps` can instead be printed if point 2 can not be achieved.

The end goal should be a JSON object, similar to:
```json
[
{"file":"path/to/file1","results":[
  {"description":"","shortname":"lic1","sim_score":1,"sim_type":"TF-IDF Cosine Sim"}
]},
{"file":"path/to/file2","results":[
  {"description":"","shortname":"lic2","sim_score":1,"sim_type":"TF-IDF Cosine Sim"}
]},
{"file":"path/to/file3","results":[
  {"description":"","shortname":"lic3","sim_score":1,"sim_type":"TF-IDF Cosine Sim"}
]}
]
```

GMishx reviewed (changes requested):
Few more suggestions.

SinghShreya05 commented:
1. Corrected Indentation from 4 to 2 spaces.
2. Fixed printing of output of each scan as the result comes.
3. Fixed formatting of JSON output to get the desired format.

Kindly review it @GMishx

GMishx commented on atarashi/atarashii.py:
```diff
@@ -126,28 +126,58 @@ def main():
+              }]
+            result = list(result)
+            result = {"file": fpath, "results": result}  
+            for k,v in result.items():
+              print(k," : ",v)          
+            dirresult.append(result)
+      dirresult = json.dumps(dirresult, sort_keys=True, ensure_ascii=False, indent=4)                
```
The `for k,v` loop is not required if you are appending results to `dirresult`. A single print of the `dirresult` should be good enough at the end.

GMishx commented:
Tried running the code, got following error:
```
$ atarashi -a Ngram -s BigramCosineSim atarashi/libs/
Traceback (most recent call last):
  File "/atarashi/.venv/bin/atarashi", line 33, in <module>
    sys.exit(load_entry_point('atarashi', 'console_scripts', 'atarashi')())
  File "/atarashi/atarashi/atarashii.py", line 153, in main
    for dirpath, dirnames, filenames in os.walk(filepath):
NameError: name 'filepath' is not defined
```

GMishx commented on atarashi/atarashii.py:
```diff
@@ -126,28 +126,57 @@ def main():
+            result = list(result)
+            result = {"file": fpath, "results": result}  
+            dirresult.append(result)
+      print(dirresult)
+      dirresult = json.dumps(dirresult, sort_keys=True, ensure_ascii=False, indent=4)                
```
I would rather print it after `json.dumps` operation to get a valid JSON. What do you think?

SinghShreya05 replied:
Actually, to keep the user interface interactive, I thought of printing "for loop consisting of k,v results.items" while the "elif" loop runs. So that simultaneously user can view the filename and its results of each file scan as and when it comes. The final result will be dumped as JSON in dirresult. 
Do you want me to plain print "dirresult" after json.dumps?

GMishx replied:
It will be best to print it in interactive way but keeping the JSON syntax valid is also a must. If you can understand C++ code, you can refer to these bit
here: https://github.com/fossology/fossology/blob/master/src/ojo/agent/OjoUtils.cc#L365
and here: https://github.com/fossology/fossology/blob/master/src/ojo/agent/directoryScan.cc#L55

What it does is, it prints a `[` at the beginning of the loop and then
1. Stringify a single JSON object.
2. If it is a first object, do not print a `,` otherwise print it (maintained by `printComma`).
3. Print the object string.

When the loop ends, print the last `]`

This is one of the way I implemented it in the CPP code. If there is a better way of doing it, please do implement that instead.

SinghShreya05 commented:
1. Implemented json.dumps() instead of simple print.
2. The array generated contains "," between two objects.

SinghShreya05 commented:
1. Removed "dirresult" variable from the main function.
2. Changed the argparse input "inputFile" to "inputPath" and the help message as well.
Please review it.

GMishx reviewed (approved):
Changes looks good.

GMishx reviewed (approved):
Tested, working as expected.

GMishx commented:
@SinghShreya05 , since we try to keep our git history clean and this PR adds a single feature (which can not be divided into multiple commits), can you please squash the commits into single commit before we can merge it?

SinghShreya05 commented:
@GMishx I have created a new branch "feat/scandirectory" and sent a new pull request with just one commit. I'm closing this PR.

Kaushl2208 commented:
@SinghShreya05, That's not how this works. He will need to review and test the functionality again. It will cost more time and energy. Try squashing the commits for the PR which is approved.
